### PR TITLE
Make `tasks/incrementalBuild-customTaskClass` snippets compatible with configuration cache

### LINKED ISSUE #21820: CCH22-23: Adjust the `snippets/tasks/incrementalBuild-customTaskClass` sample for configuration cache
The sample project located under [`subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass`](https://github.com/gradle/gradle/tree/master/subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass) does not meet [the configuration cache requirements](https://docs.gradle.org/nightly/userguide/configuration_cache.html#config_cache:requirements).

The build scripts in the `groovy` and `kotlin` parts of the sample project need to be adjusted to follow the requirements.
### Context
This issue is a part of the [Gradle Configuration Cache Hackathon 2022](https://github.com/gradle/cc-hackathon-2022).

If you wish to participate and contribute the fix, please follow [the suggested workflow](https://github.com/gradle/cc-hackathon-2022/blob/main/workflow.md).
### Expected behavior

- The sample project is fixed so that the related documentation tests pass with configuration cache enabled.
- The tests related to this sample project are removed from the `testsToBeFixedForConfigurationCache` list in `subprojects/docs/build.gradle`. You can find the test IDs below.
- The doc pages that include the snippets from this sample project might need to be updated to be consistent with the sample. Feel free to ask the team for help with this! You can find the pages listed below.
### Test failure details
These tests running in `:docs:docsTest` check the sample project and are failing with configuration cache enabled:
* `snippet-tasks-incremental-build-custom-task-class_groovy_customTaskClassWithInputOutputAnnotations.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_groovy_customTaskClassWithInputOutputAnnotations.sample)
    ```
    Task `:processTemplates` of type `org.example.ProcessTemplates`: invocation of 'Task.project' at execution time is unsupported.
    ```

* `snippet-tasks-incremental-build-custom-task-class_groovy_incrementalAdHocTask.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_groovy_incrementalAdHocTask.sample)
    ```
    Could not find method copy() for arguments [build_5yks5tcxomo6e692nd30tvva1$_run_closure2$_closure9$_closure10@2b7489cb] on task ':processTemplatesAdHoc' of type org.gradle.api.DefaultTask.
    ```

* `snippet-tasks-incremental-build-custom-task-class_groovy_inferredTaskDep2.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_groovy_inferredTaskDep2.sample)
    ```
    Task `:processTemplates` of type `org.example.ProcessTemplates`: invocation of 'Task.project' at execution time is unsupported.
    ```

* `snippet-tasks-incremental-build-custom-task-class_groovy_inferredTaskDep.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_groovy_inferredTaskDep.sample)
    ```
    Task `:packageFiles` of type `org.gradle.api.tasks.bundling.Zip`: cannot serialize object of type 'org.example.ProcessTemplates', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
    Task `:processTemplates` of type `org.example.ProcessTemplates`: invocation of 'Task.project' at execution time is unsupported.
    ```

* `snippet-tasks-incremental-build-custom-task-class_kotlin_customTaskClassWithInputOutputAnnotations.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_kotlin_customTaskClassWithInputOutputAnnotations.sample)
    ```
    Task `:processTemplates` of type `org.example.ProcessTemplates`: invocation of 'Task.project' at execution time is unsupported.
    ```

* `snippet-tasks-incremental-build-custom-task-class_kotlin_incrementalAdHocTaskNoSource.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_kotlin_incrementalAdHocTaskNoSource.sample)
    ```
    Task `:processTemplatesAdHocSkipWhenEmpty` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
    ```

* `snippet-tasks-incremental-build-custom-task-class_kotlin_incrementalAdHocTask.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_kotlin_incrementalAdHocTask.sample)
    ```
    Task `:processTemplatesAdHoc` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
    ```

* `snippet-tasks-incremental-build-custom-task-class_kotlin_inferredTaskDep2.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_kotlin_inferredTaskDep2.sample)
    ```
    Task `:processTemplates` of type `org.example.ProcessTemplates`: invocation of 'Task.project' at execution time is unsupported.
    ```

* `snippet-tasks-incremental-build-custom-task-class_kotlin_inferredTaskDep.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-build-custom-task-class_kotlin_inferredTaskDep.sample)
    ```
    Task `:packageFiles` of type `org.gradle.api.tasks.bundling.Zip`: cannot serialize object of type 'org.example.ProcessTemplates', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
    Task `:processTemplates` of type `org.example.ProcessTemplates`: invocation of 'Task.project' at execution time is unsupported.
    ```

### How to reproduce
To reproduce the failure in the Gradle documentation tests, run the following in the `gradle/gradle`repository:
```sh
./gradlew :docs:docsTest \
    -PrunBrokenConfigurationCacheDocsTests=true \
    -PenableConfigurationCacheForDocsTests=true \
    --tests "*snippet-tasks-incremental-build-custom-task-class_groovy_customTaskClassWithInputOutputAnnotations.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_groovy_incrementalAdHocTask.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_groovy_inferredTaskDep2.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_groovy_inferredTaskDep.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_kotlin_customTaskClassWithInputOutputAnnotations.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_kotlin_incrementalAdHocTaskNoSource.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_kotlin_incrementalAdHocTask.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_kotlin_inferredTaskDep2.sample" \
    --tests "*snippet-tasks-incremental-build-custom-task-class_kotlin_inferredTaskDep.sample"
```
In the test build outputs, which you can find in the test report, the configuration cache problems are listed along with the URL pointing to the detailed HTML problems report that can help investigating the violation. 
You can find more instructions on working with the sample projects in the [FAQ](https://github.com/gradle/cc-hackathon-2022/blob/main/faq.md).
### Docs including the snippets from this sample project
The following doc pages include the snippets from this sample project: 
- `subprojects/docs/src/docs/userguide/running-builds/performance.adoc`;
- `subprojects/docs/src/docs/userguide/authoring-builds/more_about_tasks.adoc`;

### PR DESCRIPTION
**Pull request by vmmaldonadoz**
Fixes #21820

### Contributor Checklist
- [X] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [X] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [X] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [X] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [X] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [X] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [X] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [X] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [X] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle [hunk 1]
```diff
@@ -87,7 +101,7 @@ tasks.register('processTemplatesWithExtraInputs', ProcessTemplates) {
 
 // tag::inferred-task-dep-via-outputs[]
 tasks.register('packageFiles', Zip) {
-    from processTemplates.map {it.outputs }
+    from processTemplates.map { it.outputDir }
```
Comment (vmmaldonadoz) [inline subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle:104]: This change was made to reference a supported serializable object, maintaining the implicit dependency between tasks.

File: subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle [hunk 2]
```diff
@@ -23,12 +27,18 @@ tasks.register('processTemplatesAdHoc') {
     outputs.dir(layout.buildDirectory.dir('genOutput2'))
         .withPropertyName('outputDir')
 
+// end::ad-hoc-task[]
```
Comment (vmmaldonadoz) [inline subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle:30]: Added a few tags for hiding details that are not related to the documentation section.

File: subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/tests/customTaskClassWithInputOutputAnnotationsUpToDate.out [hunk 3]
```diff
@@ -1,4 +1,4 @@
 > Task :processTemplates UP-TO-DATE
 
 BUILD SUCCESSFUL in 0s
-3 actionable tasks: 3 up-to-date
+1 actionable task: 1 up-to-date
```
Comment (vmmaldonadoz) [inline subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/tests/customTaskClassWithInputOutputAnnotationsUpToDate.out:4]: Here, I updated the expected outputs for the test [customTaskClassWithInputOutputAnnotations.sample.conf](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/tests/customTaskClassWithInputOutputAnnotations.sample.conf).

This change was required because in this test the task `processTemplates` has to be executed twice. 
Without configuration cache feature, two additional tasks coming from [buildSrc](https://scans.gradle.com/s/foqi3mkyjwzty/timeline?hide-timeline&outcome=from_cache,up_to_date&sort=longest) were executed along with `processTemplates`. 
However, with configuration cache, in the first execution of `processTemplates` the outputs from the `buildSrc`'s tasks are cached too and don't appear as executed (or up-to-date) in the second run.

**Comment by abstratt** [comment]
@vmmaldonadoz Running the snippet tests failed with your changes. Note that to run tests *after* you removed the failing snippets from `subprojects/docs/build.gradle`, you must no longer use the `-PrunBrokenConfigurationCacheDocsTests=true` property (or set it to `false` instead).

The main issue has to do with the ProcessTemplates class, which will need to be changed as well in order to be compatible with the configuration cache.

**Review by abstratt — CHANGES_REQUESTED** [review]
Please see feedback

**Comment by abstratt** [comment]
@bot-gradle test BD

**Comment by abstratt** [comment]
@bot-gradle test this

**Comment by vmmaldonadoz** [comment]
> @vmmaldonadoz Running the snippet tests failed with your changes. Note that to run tests _after_ you removed the failing snippets from `subprojects/docs/build.gradle`, you must no longer use the `-PrunBrokenConfigurationCacheDocsTests=true` property (or set it to `false` instead).
> 
> The main issue has to do with the ProcessTemplates class, which will need to be changed as well in order to be compatible with the configuration cache.
> 
> Also, to reproduce issues by running the project alone:
> 
> ```
> cd subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy
> ../../../../../../../gradlew processTemplates --rerun-tasks --configuration-cache
> ```
> 
> Both `--configuration-cache` and `--rerun-tasks` are required to expose the problem.
> 
> Please be mindful that I sync'd your branch with master.

Hey @abstratt. You were right about the required changes in  `ProcessTemplates.java` (somehow I didn't commit the changes and the tests passed locally). Thank you for the insights about the issue 👍

**Comment by abstratt** [comment]
@bot-gradle BD

**Comment by abstratt** [comment]
@bot-gradle test BD RFN

File: subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/tests/customTaskClassWithInputOutputAnnotationsUpToDate.out (hunk 3 above)
Comment (abstratt) [reply subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/tests/customTaskClassWithInputOutputAnnotationsUpToDate.out:4]: Sorry for the long delay on responding to this, @vmmaldonadoz. This poses a challenge to our testing framework, as there is no single expected output that will work with and without the configuration cache. We will are discussing what we could do here - disabling CC for the `processTemplates` task would be a possible workaround.

**Review by abstratt — APPROVED** [review]

**Comment by abstratt** [comment]
@bot-gradle  test this

**Comment by abstratt** [comment]
@bot-gradle cancel

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
Just a network failure. @bot-gradle test and merge

**Comment by abstratt** [comment]
🎉
